plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/file/PomFileAnnotator.kt

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codemodernizer.file
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.actionSystem.ActionGroup
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.application.runReadAction
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.impl.DocumentMarkupModel.forDocument
+import com.intellij.openapi.editor.markup.EffectType
+import com.intellij.openapi.editor.markup.GutterIconRenderer
+import com.intellij.openapi.editor.markup.HighlighterLayer
+import com.intellij.openapi.editor.markup.HighlighterTargetArea
+import com.intellij.openapi.editor.markup.MarkupModel
+import com.intellij.openapi.editor.markup.TextAttributes
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.fileEditor.OpenFileDescriptor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.ui.JBColor
+import software.aws.toolkits.jetbrains.services.codemodernizer.constants.CodeModernizerUIConstants.Companion.getLightYellowThemeBackgroundColor
+import software.aws.toolkits.resources.AwsToolkitBundle.message
+import java.awt.Color
+import java.awt.Font
+import javax.swing.Icon
+
+class PomFileAnnotator(private val project: Project, private var virtualFile: VirtualFile, private var lineNumberToHighlight: Int?) {
+    val emptyAction: AnAction = object : AnAction() {
+        override fun actionPerformed(e: AnActionEvent) = Unit
+        override fun update(e: AnActionEvent) = Unit
+    }
+
+    // If corresponding model doesn't exist, create it for that document
+    private fun getMarkupModelForDocument(document: Document): MarkupModel =
+        forDocument(document, project, false) ?: forDocument(document, project, true)
+
+    private fun openVirtualFile() {
+        val fileEditorManager = FileEditorManager.getInstance(project)
+        val openFileDescription = OpenFileDescriptor(project, virtualFile)
+        fileEditorManager.openTextEditor(openFileDescription, true)
+    }
+
+    private fun addGutterIconToLine(markupModel: MarkupModel, document: Document, lineNumberToHighlight: Int) {
+        val gutterIconRenderer = object : GutterIconRenderer() {
+            override fun equals(other: Any?): Boolean = true
```

Comment:
? don't think this should always be true

Reply:
Since we have just one and same icon, rendering just once seems to be fine.

Reply:
Converted to  a private class.